lib/core/src/Cardano/Wallet/Primitive/AddressDerivation/Random.hs

```diff
@@ -77,6 +89,33 @@ import qualified Data.ByteArray as BA
 import qualified Data.ByteString as BS
 import qualified Data.ByteString.Lazy as BL
 
+{-------------------------------------------------------------------------------
+                                   Key Types
+-------------------------------------------------------------------------------}
+
+-- | Data for generating random scheme addresses
+data RndKey (level :: Depth) key = RndKey
+    { rootKey :: Key 'RootK XPrv
+    , addrKey :: Key 'AddressK XPub -- address public key
```

Comment:
Shouldn't this second field be a `Key 'level key` :thinking: ?
We only have an address key for the `AddressK` depth, and it can be public or private. So in practice we do manipulate:

```
RndKey 'RootK XPrv
RndKey 'RootK XPub
RndKey 'AccountK XPrv
RndKey 'AddressK XPrv
RndKey 'AddressK XPub
```

We do always have a root key, but we only have an account index for keys below the `AccountK` depth, and only an address index for keys at `AddressK` depth.

Reply:
If I understand it properly, we need 
`Key 'RootK XPrv` (or alternatively seed and encryption passwords) AND
`encryption password`, `account index` and `address index`.
With those we can derive key for any address. This is done : 
```
rootXPrv = generateKeyFromSeed seed encPwd
```
in order to get root private key that of type `Key 'RootK XPrv`
Knowing that we can get : 
```   accXPrv = deriveAccountPrivateKey encPwd rootXPrv accIx
    addrXPrv = deriveAddressPrivateKey encPwd accXPrv addrIx
```
or public counterparts by applying `publicKey`.
Saying that can't `RndKey` be : 
```   
data RndKey (level :: Depth) key = RndKey
    { rootKey :: Key 'RootK XPrv
    , encryptionPassword :: Password "encryption" 
    , accountIndex :: Index 'Hardened 'AccountK
    , addressIndex :: Index 'Soft 'AddressK
    } deriving stock (Generic)
```
?

Reply:
@paweljakubas key derivation is a rather expensive operation that ideally we don't want to be re-performing all the time so we would need to at least store the cryptographic content inside the `RndKey` (what we currently wrap in a `Key (* :: Depth) *`. 

Having said that, note that we either need the root public key, or the already computed payload passphrase (which is a different thing than the master encryption passphrase!) which is specific to the rnd scheme and can be computed from the root public key.

Reply:
Thanks for the help with this guys. Please check out the new version of `RndKey`. It does now contain enough information to implement `rndToAddress`. But it's sort of weird because it has a `level` type parameter which is not used in any of the fields.